### DISCUSSION TITLE
Hawtio Roadmap 2024

### DISCUSSION THREAD

**Discussion by tadayosi** [hawtio/hawtio#3400] [Announcements]
About two weeks ago we released [Hawtio 4.0.0 GA](https://github.com/hawtio/hawtio/releases/tag/hawtio-4.0.0). 4.0.0 is mainly a [release aimed at supporting new technology stacks](https://github.com/orgs/hawtio/discussions/3185), which has made Hawtio fully available with the modern Java ecosystem (Jakarta EE 9+, Quarkus 3, Spring Boot 3, and Apache Camel 4).

Now the project can look to the future and start working on new features. We are already a quarter of the way through 2024, and I'd like to share the themes and new features that the team is planning to work on this year.

**Common:**

- Continuous improvement of Hawtio based on community feedback
- PatternFly upgrade to v5 / v6
- Restore Dashboard feature
- Integration with Kaoto at component level
- Implement RBAC support for Quarkus and Spring Boot
- Flight Recorder plugin (a.k.a. Diagnostics plugin)
- Exploring new protocols alternative to Jolokia/JMX

**Hawtio Online:**

- Performance improvement
- Integrate more into OpenShift

## Continuous improvement of Hawtio based on community feedback

We already have users who've been trying out the new Hawtio since the 3.0 GA and 4.0 beta versions. We have also already received reports of bugs and feature improvements from the users. We will continue to address these existing reports that were not incorporated into the 4.0.0 GA, as well as the following reports we will continue to accept:

- New features/enhancements
- UX improvements
- Bug fixes

## PatternFly upgrade to v5 / v6

- https://github.com/hawtio/hawtio-next/issues/477

When we started developing Hawtio v3, the latest version of PatternFly was still v4. Therefore, we decided to use PatternFly v4 as the UI base until the GA release of Hawtio v3 and v4. However, the latest PatternFly is already v5 and v6 release is planned for later this year.

Therefore, as a major initiative immediately after the release of Hawtio 4.0.0, we plan to migrate to PatternFly v5 and have already started work on this.

We will also start looking at migrating to v6 later in the year.

## Restore Dashboard feature

- https://github.com/hawtio/hawtio-next/issues/541

The good old Hawtio v1 had an eye-catching feature: Dashboard. However, in the process of moving to PatternFly with Hawtio v2, the feature was temporarily discontinued as there was no corresponding feature in PatternFly v3 at the time.

Now that Hawtio v4 has matured as a management console, and as we look to further increase integration into Kubernetes/OpenShift through Hawtio Online, we are looking to bring back the dashboard functionality again.

To implement the dashboard, the part of the React UI components implemented in [hawtio-next](https://github.com/hawtio/hawtio-next) needs to be improved as follows:

- Make the plugins more thoroughly reusable React components
- Make the Hawtio UI part frameless so that the plugins can be run on their own (headless mode)

## Integration with Kaoto at component level

[Kaoto](https://github.com/KaotoIO/kaoto) is a sister project to Hawtio as a visual tooling for Apache Camel. Whereas Hawtio provides _runtime_ visual views for Camel routes, Kaoto is a _design-time_ visual editor for them. Despite the differences, they share many common features in terms of graphically displaying and editing Camel routes.

Therefore, in order to accelerate the evolution of the two projects together, we will explore the possibility of sharing functionality at component level. Sharing the Camel editor at component level would offer many benefits to the users, such as common functionality and look and feel between the two tools.

## Implement RBAC support for Quarkus and Spring Boot

- https://github.com/hawtio/hawtio/issues/3045

RBAC was one of the outstanding features of Hawtio v2, but was only offered for limited runtimes (Apache Karaf, Apache ActiveMQ Artemis, and Hawtio Online) due to mandatory backend support. Even in Hawtio v4, RBAC is currently only supported by Hawtio Online and Apache ActiveMQ Artemis.

Extending this RBAC support to the primary runtimes, Quarkus and Spring Boot, will be one of the major goals this year.

## Flight Recorder plugin (a.k.a. Diagnostics plugin)

- https://github.com/hawtio/hawtio-next/issues/643

The Diagnostics plugin was an excellent plugin that @skarsaune contributed to Hawtio v2. Unfortunately we haven't had time to port it to Hawtio v4 so far. But, as it would be too good to discontinue it as it is, we are planning to port it to v4.

## Exploring new protocols alternative to JMX/Jolokia

- https://github.com/hawtio/hawtio-next/issues/170

Hawtio v4 fully supports Quarkus along with Spring Boot as runtime. However, it does not provide a Quarkus plugin equivalent to [Spring Boot plugin](https://github.com/hawtio/hawtio-next/tree/main/packages/hawtio/src/plugins/springboot) at the Hawtio UI level. We haven't been able to provide a Quarkus plugin, because Quarkus is a framework that assumes native Java builds and thus, unlike Spring Boot, does not provide standard MBeans to manage applications via JMX which has less affinity with native builds.

In order to provide a UI plugin equivalent to Spring Boot for Quarkus, support for a protocol different from JMX/Jolokia should be considered. As applications become more cloud-native, more and more frameworks are providing health check endpoints etc. that do not rely on JMX. It is therefore worth considering support for protocols other than JMX for the future of Hawtio.

## Hawtio Online: Performance improvement

- https://github.com/hawtio/hawtio/issues/3182

While the original issue is not specific to Hawtio Online, reports have been raised of performance issues when viewing fairly large Camel applications or Artemis brokers with many destinations on the Hawtio console. It is one of Hawtio's goals to be able to run nimbly at the scale of such practical level applications. In particular, Hawtio Online is prone to these scaling-related performance impacts due to differences in deployment architecture.

Therefore, while we work on performance tuning of Hawtio as a whole, this year we will primarily focus on optimising Hawtio Online.

- https://github.com/hawtio/hawtio-next/issues/754

Furthermore, Hawtio currently [supports multiple LTS versions of the Camel model](https://github.com/hawtio/hawtio-next/issues/693). This requires the inclusion of two or more Camel model definitions in the Hawtio UI bundle, and its large definition files are downloaded on page load, resulting in very slow initial loading of the Hawtio console. This is a particularly noticeable problem with Hawtio Online deployed on a remote cluster.

Making the Camel models lazy loaded should solve the performance problem during initial page loading of Hawtio Online.

## Hawtio Online: Integrate more into OpenShift

- https://github.com/hawtio/hawtio-online/issues/133

Hawtio Online already integrates Hawtio with OpenShift/Kubernetes by providing the [Operator](https://github.com/hawtio/hawtio-operator), but we are aiming for further integration and more cluster-native user experiences.

One of the initiatives is to allow Hawtio Online to be installed as a [dynamic plugin](https://docs.okd.io/latest/web_console/dynamic-plugin/overview-dynamic-plugin.html) for OpenShift. If realised, this would provide three ways to deploy Hawtio Online to OpenShift:

- CRD-based manual installation
- Operator-based installation
- Installation as a dynamic plugin

**Comment by grgrzybek** [comment]
Hello

Great and amazing summary!

I assigned myself to:
* FlightRecording one: https://github.com/hawtio/hawtio-next/issues/643
* Implement RBAC support for Quarkus and Spring Boot:
   https://github.com/hawtio/hawtio/issues/3045
* New protocols: https://github.com/hawtio/hawtio-next/issues/170

**Reply by tadayosi** [reply]
Thank you!

While I wrote as "new protocols alternative to JMX/Jolokia", as you suggest supporting new protocols within Jolokia is also a great idea to explore. That would also keep Jolokia relevant and active in the future.

**Comment by mmelko** [comment]
I assigned myself to Patternfly migration, all-tough we should discuss further we we won't migrate directly to 6 if the release is so close. I can also work on Kaoto+Hawtio integration on the component level, especially sharing the visualisation part.
